File: Stack/0020. Valid Parentheses/0020.cpp
class Solution {
public:
    bool isValid(string s) {
        std::stack<char> myStack;
        
        for (size_t i = 0; i < s.length(); ++i) {
            if (s[i] == '(' || s[i] == '[' || s[i] == '{') {
                myStack.push(s[i]);
            }
                
            // s[i] == closing bracket or paraenthesis
            else {
                if (!myStack.empty()) {
                    if ((s[i] == ')' && myStack.top() == '(') 
                    || (s[i] == ']' && myStack.top() == '[') 
                    || (s[i] == '}' && myStack.top() == '{')) {
                        myStack.pop();
                    }
                    // s[i] does not match myStack.top() --> imbalanced string
                    else {
                        return false;
                    }
                }
                // stack is empty --> imbalanced string
                else {
                    return false;
                }
            }
        }

        if (myStack.empty()) {
            return true;
        }
        return false;
    }
};

File: Stack/0020. Valid Parentheses/0020.py
class Solution:
    def isValid(self, s: str) -> bool:
        CloseToOpen = {")": "(", "]": "[", "}": "{"}
        stack = []
 
        for c in s:
            if c in CloseToOpen: # checking the key, thus checking if c is a closing parenthesis
                if stack and stack[-1] == CloseToOpen[c]: # if stack is not empty && last open parenthesis matches closing parenthesis
                    stack.pop()
                else:
                    return False
            else: # NOT a closing paranthesis
                stack.append(c)

        return True if not stack else False # only return true if stack is empty else return False
